### PR TITLE
team/hier: propagate net devices to inner teams

### PR DESCRIPTION
**Pull request by vspetrov**

### CODE REVIEW

File: src/team_lib/hier/xccl_hier_context.c [hunk 1]
```diff
@@ -34,24 +33,24 @@ xccl_hier_init_tl(xccl_hier_context_t *ctx, xccl_tl_id_t tl_id,
...
-                 prefix, xccl_tl_str(1 << tl_id));
+                 hier_cfg->super.env_prefix, xccl_tl_str(tl_id));
     }
     else {
-        snprintf(env_prefix, sizeof(env_prefix), "HIER_%s", xccl_tl_str(1 << tl_id));
+        snprintf(env_prefix, sizeof(env_prefix), "HIER_%s", xccl_tl_str(tl_id));
     }
-
     xccl_context_config_read(lib, env_prefix, NULL, &cfg);
+    if (hier_cfg->devices.count > 1 ||
```
Comment (Sergei-Lebedev) [inline src/team_lib/hier/xccl_hier_context.c:48]: Maybe we should allow to pass 'all' as net_devices? UCX recognizes this as "all available devices"

File: src/team_lib/hier/xccl_hier_context.c [hunk 2]
```diff
@@ -117,12 +116,24 @@ xccl_status_t xccl_hier_create_context(xccl_team_lib_t *lib,
...
     int                 i;
     uint64_t            tl;
-
     XCCL_CONTEXT_SUPER_INIT(ctx->super, lib, params);
+
+    if (NULL == hlib->tl_lib) {
+        xccl_lib_params_t lib_params = {
+            .field_mask = XCCL_LIB_PARAM_FIELD_TEAM_USAGE,
+            .team_usage = XCCL_LIB_PARAMS_TEAM_USAGE_SW_COLLECTIVES |
+            XCCL_LIB_PARAMS_TEAM_USAGE_HW_COLLECTIVES,
```
Comment (Sergei-Lebedev) [inline src/team_lib/hier/xccl_hier_context.c]: alignment

File: src/team_lib/hier/xccl_hier_context.c [hunk 3]
```diff
@@ -34,24 +33,24 @@ xccl_hier_init_tl(xccl_hier_context_t *ctx, xccl_tl_id_t tl_id,
...
     else {
-        snprintf(env_prefix, sizeof(env_prefix), "HIER_%s", xccl_tl_str(1 << tl_id));
+        snprintf(env_prefix, sizeof(env_prefix), "HIER_%s", xccl_tl_str(tl_id));
     }
-
     xccl_context_config_read(lib, env_prefix, NULL, &cfg);
+    if (hier_cfg->devices.count > 1 ||
+        0 != strncmp(hier_cfg->devices.names[0], "all", 3)) {
+        xccl_context_config_modify(&tl_id, cfg, "NET_DEVICES",
+                                   xccl_names_array_to_str(&hier_cfg->devices));
```
Comment (Sergei-Lebedev) [inline src/team_lib/hier/xccl_hier_context.c]: Need to free hier_cfg->devices?

File: src/team_lib/hier/xccl_hier_context.c [hunk 4]
```diff
@@ -117,12 +116,24 @@ xccl_status_t xccl_hier_create_context(xccl_team_lib_t *lib,
...
     XCCL_CONTEXT_SUPER_INIT(ctx->super, lib, params);
+
+    if (NULL == hlib->tl_lib) {
+        xccl_lib_params_t lib_params = {
+            .field_mask = XCCL_LIB_PARAM_FIELD_TEAM_USAGE,
+            .team_usage = XCCL_LIB_PARAMS_TEAM_USAGE_SW_COLLECTIVES |
+            XCCL_LIB_PARAMS_TEAM_USAGE_HW_COLLECTIVES,
+        };
+        if (XCCL_OK != xccl_lib_init(&lib_params, NULL, &hlib->tl_lib)) {
+            return XCCL_ERR_NO_MESSAGE;
```
Comment (Sergei-Lebedev) [inline src/team_lib/hier/xccl_hier_context.c]: return actual error instead of XCCL_ERR_NO_MESSAGE

File: src/team_lib/hier/xccl_hier_context.c [hunk 5]
```diff
@@ -147,7 +158,7 @@ xccl_status_t xccl_hier_create_context(xccl_team_lib_t *lib,
     ctx->bcast_sm_get_thresh                    = hier_cfg->bcast_sm_get_thresh;
 
     ucs_for_each_bit(tl, XCCL_TL_ALL) {
-        if (XCCL_OK != xccl_hier_init_tl(ctx, tl, params->oob, config->env_prefix)) {
+        if (XCCL_OK != xccl_hier_init_tl(ctx, tl, params->oob, hier_cfg)) {
             return XCCL_ERR_NO_MESSAGE;
```
Comment (Sergei-Lebedev) [inline src/team_lib/hier/xccl_hier_context.c]: return actual error instead of XCCL_ERR_NO_MESSAGE

File: src/team_lib/hier/xccl_hier_context.c [hunk 6]
```diff
@@ -34,24 +33,24 @@ xccl_hier_init_tl(xccl_hier_context_t *ctx, xccl_tl_id_t tl_id,
...
     xccl_context_config_read(lib, env_prefix, NULL, &cfg);
+    if (hier_cfg->devices.count > 1 ||
+        0 != strncmp(hier_cfg->devices.names[0], "all", 3)) {
+        xccl_context_config_modify(&tl_id, cfg, "NET_DEVICES",
+                                   xccl_names_array_to_str(&hier_cfg->devices));
+    }
     if (XCCL_OK != xccl_context_create(lib, &ctx_params, cfg,
-                                       &ctx->tls[tl_id].xccl_ctx)) {
+                                       &ctx->tls[tl_idx].xccl_ctx)) {
         return XCCL_ERR_NO_MESSAGE;
```
Comment (Sergei-Lebedev) [inline src/team_lib/hier/xccl_hier_context.c]: return actual error instead of XCCL_ERR_NO_MESSAGE

File: src/utils/utils.h [hunk 7]
```diff
@@ -38,4 +38,26 @@ unsigned long xccl_str_hash(const char *str) {
...
+    if (!str) {
+        return NULL;
+    }
+    strcpy(str, array->names[0]);
+    for (i=1; i<array->count; i++) {
+        strcat(str, ",");
+        strcat(str, array->names[i]);
+    }
+    return str;
+}
```
Comment (Sergei-Lebedev) [inline src/utils/utils.h:62]: new line before #endif

File: src/team_lib/hier/xccl_hier_context.c [hunk 8]
```diff
@@ -117,12 +116,24 @@ xccl_status_t xccl_hier_create_context(xccl_team_lib_t *lib,
                                        xccl_tl_context_config_t *config,
                                        xccl_tl_context_t **context)
 {
-    xccl_hier_context_t *ctx = (xccl_hier_context_t *)malloc(sizeof(*ctx));
+    xccl_team_lib_hier_t *hlib = ucs_derived_of(lib, xccl_team_lib_hier_t);
+    xccl_hier_context_t  *ctx  = (xccl_hier_context_t *)malloc(sizeof(*ctx));
```
Comment (Sergei-Lebedev) [inline src/team_lib/hier/xccl_hier_context.c:120]: check for NULL and return ERR_NO_MEMORY

**Review by Sergei-Lebedev — CHANGES_REQUESTED** [review]

File: src/team_lib/hier/xccl_hier_context.c (hunk 1 above)
Comment (vspetrov) [reply src/team_lib/hier/xccl_hier_context.c:48]: The idea is:  if user does not specify XCCL_HIER_NET_DEVICES then we should not overwrite potential HIER_ucx_XCCL_TEAM_UCX_NET_DEVICES with "all".

File: src/team_lib/hier/xccl_hier_context.c (hunk 3 above)
Comment (vspetrov) [reply src/team_lib/hier/xccl_hier_context.c]: 👍

File: src/team_lib/hier/xccl_hier_context.c (hunk 5 above)
Comment (vspetrov) [reply src/team_lib/hier/xccl_hier_context.c]: done

File: src/team_lib/hier/xccl_hier_context.c (hunk 6 above)
Comment (vspetrov) [reply src/team_lib/hier/xccl_hier_context.c]: done

File: src/team_lib/hier/xccl_hier_context.c (hunk 8 above)
Comment (vspetrov) [reply src/team_lib/hier/xccl_hier_context.c:120]: done

**Review by Sergei-Lebedev — APPROVED** [review]

**Pull request merged**